components/feature/downloads/src/test/java/mozilla/components/feature/downloads/share/ShareDownloadMiddlewareTest.kt

```diff
@@ -0,0 +1,275 @@
+        val download = DownloadState("testDownload", source = DownloadState.Source.DOWNLOAD)
+        val addDownloadAction = DownloadAction.AddDownloadAction(download)
+        val removeDownloadAction = DownloadAction.RemoveDownloadAction("")
+        val removeAllDownloadsAction = DownloadAction.RemoveAllDownloadsAction
+        val updateDownloadAction = DownloadAction.UpdateDownloadAction(mock())
+        val restoreDownloadsAction = DownloadAction.RestoreDownloadsStateAction
+        val restoreDownloadAction = DownloadAction.RestoreDownloadStateAction(mock())
+
+        // dispatch all 6 DownloadAction
+        store.dispatch(addDownloadAction).joinBlocking()
+        store.dispatch(removeAllDownloadsAction).joinBlocking()
+        store.dispatch(removeDownloadAction).joinBlocking()
+        store.dispatch(updateDownloadAction).joinBlocking()
+        store.dispatch(restoreDownloadAction).joinBlocking()
+        store.dispatch(restoreDownloadsAction).joinBlocking()
+        testDispatcher.advanceUntilIdle()
+
+        verify(shareMiddleware, never()).startSharing(any())
+        // 6 actions dispatched from above + the default InitAction
+        verify(fakeMiddleware, times(7)).invoke(any(), any(), any())
+    }
+
+    @Test
+    @Ignore("Call not recorded: \"zero interactions with this mock.\"")
+    fun `ShareDownloadMiddleware should call cleanupCache() in init()`() = runBlocking {
+        val shareMiddleware = spy(ShareDownloadMiddleware(mock(), mock(), TestCoroutineScope()))
+
+        verify(shareMiddleware).cleanupCache()
+    }
+
+    @Test
+    fun `cleanupCache should delete all files from the cache directory`() = runBlocking {
+        val shareMiddleware = spy(ShareDownloadMiddleware(mock(), mock()))
+        val testDir = File("testDir").also { dir ->
+            dir.mkdirs()
+            File(dir, "testFile").also { file ->
+                file.createNewFile()
+            }
+        }
+
+        doReturn(testDir).`when`(shareMiddleware).getCacheDirectory()
+        assertTrue(testDir.listFiles()!!.isNotEmpty())
+
+        shareMiddleware.cleanupCache()
+
+        assertTrue(testDir.listFiles()!!.isEmpty())
+    }
+
+    @Test
+    fun `startSharing() will download and then share the selected download`() = runBlocking {
+        val context = spy(testContext)
+        val shareMiddleware = spy(ShareDownloadMiddleware(context, mock(), TestCoroutineScope()))
+        val downloadState = DownloadState("testUrl")
+        val downloadedFile = File("filePath")
+        doReturn(downloadedFile).`when`(shareMiddleware).download(any())
+
+        shareMiddleware.startSharing(downloadState)
+
+        verify(shareMiddleware).download(downloadState)
+        // The below check fails with InvalidUseOfMatchersException: 0 matchers expected, 4 recorded:
+        // verify(context).shareMedia(eq(downloadedFile.canonicalPath), eq(null), eq(null), eq("testUrl"))
```

Comment:
I think the problem is that `shareMedia` is an extension function to test it, we could add an internal function on `ShareDownloadMiddleware` that calls `context.shareMedia()` inside, and then verify that function instead:


```kotlin
// In ShareDownloadMiddleware
internal fun shareMedia() =  context .shareMedia()

// In ShareDownloadMiddlewareTest
doNothing().`when`(shareMiddleware).shareMiddleware()
verify(shareMiddleware).shareMedia()




```

Reply:
In the end the issue seemed to be the one discussed here - https://github.com/mockito/mockito-kotlin/issues/240#issuecomment-450634294 but either way I needed a proxy because the real implementation of `Context.shareMedia()` was pretty convoluted to depend on in this test.